samples/js/js-doc-snippets/src/delegatingservice.js

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2021 Lightbend Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/* This code was initialised by Akka Serverless tooling.
+ * As long as this file exists it will not be re-generated.
+ * You are free to make changes to this file.
+ */
+
+// tag::delegating-action[]
+import { Action } from "@lightbend/akkaserverless-javascript-sdk";
+import { replies } from '@lightbend/akkaserverless-javascript-sdk';
+import * as grpc from '@grpc/grpc-js'; // <1>
+
+/**
+ * Type definitions.
+ * These types have been generated based on your proto source.
+ * A TypeScript aware editor such as VS Code will be able to leverage them to provide hinting and validation.
+ * 
+ * DelegatingService; a strongly typed extension of Action derived from your proto source
+ * @typedef { import("../lib/generated/delegatingservice").DelegatingService } DelegatingService
+ */
+
+/**
+ * @type DelegatingService
+ */
+const action = new Action(
+  [
+    "com/example/delegating_service.proto",
+    "com/example/counter_api.proto" // <2>
+  ],
+  "com.example.DelegatingService",
+  {
+    includeDirs: ["./proto"],
+    serializeFallbackToJson: true
+  }
+);
+
+const counterClient = new action.grpc.com.example.CounterService(
+  "counter:80", // <3>
+  grpc.credentials.createInsecure()); // <4>
+
+// end::delegating-action[]
+function showExternal() {
+  // tag::public-grpc[]
+  const counterClient = new action.grpc.com.example.CounterService(
+    "public-host-name.akkaserverless.com",
+    grpc.credentials.createSsl());
+  // end::public-grpc[]
+}
+// tag::delegating-action[]
+
+
+action.commandHandlers = {
+  // FIXME grpc-js api-docs says there is async rpc unary method calls
+  // but actually calling without callback fails on param validation in grpc-js
+  // so for now we do this promise dance ourselves
+  async AddAndReturn(request, ctx) {
+    const increaseDone = await new Promise((resolve, reject) => {
+      counterClient.increase({counterId: request.counterId, value: 1}, // <5>
+        (error, emptyResult) => {
+          if (error) reject(error);
+          else resolve(emptyResult);
+        });
+    });
```

Comment:
I think this is the only annoyance right now, the apidoc of the service makes it look as though it can be called without any callbacks and that it will return a promise in that case, but when actually trying that out, some internal parameter validation in proto-js barfs on no callbacks passed. :/

Reply:
The integration testkit also creates clients. When adding integration tests for the samples, I also added a promisify for the client methods (originally https://github.com/lightbend/akkaserverless-framework/pull/617) so that each method has an `Async` version that returns a promise.

https://github.com/lightbend/akkaserverless-javascript-sdk/blob/227123339a17192404fd2f5b1341bb024e3e604d/sdk/src/integration-testkit.ts#L110-L124

We could do something similar here. Or rather than patching could provide our own client wrapper if we want to only have the Promise returning methods (without `Async).

Reply:
Ah, nice, so we could provide something like in the Java sdk to look up the client and return one with async-patched methods. I like that idea.